pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the "opaque ports" annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
+		if service.Spec.ClusterIP == "None" {
+			// skip headless services; they're handled differently
+			continue
+		}
+
+		endpoint, err := hc.kubeAPI.CoreV1().Endpoints(namespace).Get(ctx, service.Name, metav1.GetOptions{})
+
+		if err != nil {
+			return err
+		}
+
+		pods := make([]*corev1.Pod, 0)
+		for _, subset := range endpoint.Subsets {
+			for _, addr := range subset.Addresses {
+				if addr.TargetRef.Kind == "Pod" {
```

Comment:
I think we should check if `TargetRef` is nil first otherwise this might result in a nil ptr. dereference. I wrote a [Go playground example](https://play.golang.org/p/aJhoVAjmAzn) to illustrate this.

Basically, since `TargetRef` is a pointer to an object, accessing its attributes might panic if there is no memory address to go to (when the object is nil).

Reply:
I didn't add the check because I thought you couldn't have an `Address` with no `TargetRef`, could you?

Reply:
I think I was wrong?

```
// Reference to object providing the endpoint.
// +optional
TargetRef *ObjectReference `json:"targetRef,omitempty" protobuf:"bytes,2,opt,name=targetRef"`
```

Will add the extra check

Reply:
It may be possible to have one. This happens typically in scenarios where the Service does not include a selector and the user wants the service to point to an external IP (and/or the Endpoints object is created manually).

For example, with linkerd we have addresses with no refs in multicluster scenarios. We create our own synthetic Endpoints to point to the gateway, this would look like this:

```
apiVersion: v1
kind: Endpoints
subsets:
- addresses:
  - ip: 172.20.0.6
  ports:
  - name: mc-probe
    port: 4191
    protocol: TCP
```

I think it'd be worth having the check in just for good measure in case the namespace may include a mirrored service with the Endpoints object similar to the snippet above. Services without a selector may or may not have the target ref depending on how they were created and how they are managed.